### PR TITLE
fix(framework): prevent negative storage map balances

### PR DESCRIPTION
**Pull request by Jim8y**
## Summary
- reject negative `Increase` amounts for `StorageMap` and `LocalStorageMap`
- reject `Decrease` calls that would persist a negative value
- add framework regression tests for negative increments and underflowing decrements

## Testing
- `dotnet test tests/Neo.SmartContract.Framework.UnitTests/Neo.SmartContract.Framework.UnitTests.csproj --filter FullyQualifiedName~Neo.SmartContract.Framework.UnitTests.StorageMapSafetyTest`
- `dotnet test tests/Neo.SmartContract.Framework.UnitTests/Neo.SmartContract.Framework.UnitTests.csproj --filter "FullyQualifiedName~Neo.SmartContract.Framework.UnitTests.StorageMapSafetyTest|FullyQualifiedName~Neo.SmartContract.Framework.UnitTests.Services.StorageTest"`

Related checklist item: issue #1556, Issue 2.

### CODE REVIEW

File: src/Neo.SmartContract.Framework/Services/StorageMap.cs [hunk 1]
```diff
@@ -450,7 +452,11 @@ public BigInteger Increase(ByteString key, BigInteger amount)
         /// </summary>
         public BigInteger Decrease(ByteString key, BigInteger amount)
         {
-            var i = GetIntegerOrZero(key) - amount;
+            if (amount.Sign < 0) throw new ArgumentOutOfRangeException(nameof(amount));
+
+            var current = GetIntegerOrZero(key);
+            var i = current - amount;
+            if (i.Sign < 0) throw new InvalidOperationException();
```
Comment (shargon) [inline src/Neo.SmartContract.Framework/Services/StorageMap.cs]: For the argument is good to have this control, but for the result, I'm not sure, could you improve the error message please?

Comment (Jim8y) [reply src/Neo.SmartContract.Framework/Services/StorageMap.cs]: Updated the underflow path to throw `InvalidOperationException("result would be negative")` in both storage map helpers. `StorageMapSafetyTest` now asserts the surfaced message as well.

**Review by shargon — APPROVED** [review]

File: src/Neo.SmartContract.Framework/Services/LocalStorageMap.cs [hunk 2]
```diff
@@ -358,6 +359,7 @@ public extern ByteString? this[byte[] key]
         /// </summary>
         public BigInteger Increase(ByteString key, BigInteger amount)
         {
+            if (amount.Sign < 0) throw new ArgumentOutOfRangeException(nameof(amount));
```
Comment (Wi1l-B0t) [inline src/Neo.SmartContract.Framework/Services/LocalStorageMap.cs:362]: It's better to add 'If amount is negative, exception will be thrown` to comment.
